scientific-computing-with-python/learn-encapsulation-by-drawing-a-projectile-trajectory/projectile-trajectory.py

```diff
@@ -0,0 +1,149 @@
+        #TODO: break in smaller pieces
+        # return self.__speed * math.cos(self.__angle) * (
+        #     self.__speed * math.sin(self.__angle) +
+        #     math.sqrt(self.__speed**2 * math.sin(self.__angle)**2 +
+        #               2 * GRAVITATIONAL_ACCELERATION * self.__height)) / GRAVITATIONAL_ACCELERATION
+        horizontal_component = self.__speed * math.cos(self.__angle)
+        vertical_component = self.__speed * math.sin(self.__angle)
+        sqrt_component = math.sqrt(self.__speed ** 2 * math.sin(self.__angle) ** 2 +
+                                   2 * GRAVITATIONAL_ACCELERATION * self.__height)
+        displacement = horizontal_component * (vertical_component + sqrt_component) / GRAVITATIONAL_ACCELERATION
+        return displacement
+
+    def __calculate_y_coordinate(self, x):
+        #TODO: break in smaller pieces
+        # return self.__height + math.tan(self.__angle) * x - GRAVITATIONAL_ACCELERATION * x**2 / (
+        #     2 * self.__speed**2 * math.cos(self.__angle)**2)
+        height_component = self.__height
+        angle_component = math.tan(self.__angle) * x
+        acceleration_component = GRAVITATIONAL_ACCELERATION * x ** 2 / (
+                2 * self.__speed ** 2 * math.cos(self.__angle) ** 2)
+        y_coordinate = height_component + angle_component - acceleration_component
+        return y_coordinate
+
+    def calculate_all_coordinates(self):
+        return [
+            (x, self.__calculate_y_coordinate(x))
+            for x in range(math.ceil(self.__calculate_displacement()))
+        ]
+
+    def details(self):
+        return f'angle: {round(math.degrees(self.__angle))}°\nspeed: {self.__speed} m/s\nstarting height: {self.__height} m\ndisplacement: {round(self.__calculate_displacement(), 1)} m'
+
+    def __str__(self):
+        return self.details()
+    
+    @property
+    def height(self):
+        return self.__height
+    
+    @height.setter
+    def height(self, new_height):
+        self.__height = new_height
+
+    @property
+    def angle(self):
+        return self.__angle
+    
+    @angle.setter
+    def angle(self, new_angle):
+        self.__angle = math.radians(new_angle)
+
+    @property
+    def speed(self):
+        return self.__speed
+
+    @speed.setter
+    def speed(self, new_speed):
+        self.__speed = new_speed
+
+class Graph:
+    def __init__(self, coordinates):
+        self.__coordinates = coordinates
+
+    def __create_coordinates_table(self):
+        table = '  x      y\n'
+        for x, y in self.__coordinates:
+            table += f'{x:>3}{round(y, 2):>7.2f}\n'
+
+        return table
+
+    def __create_trajectory(self):
+        rounded_coord = [(x, round(y)) for x, y in self.__coordinates]
+        rows = max([y for x, y in rounded_coord]) + 1
+        columns = max([x for x, y in rounded_coord]) + 1
+        graph_matrix = [[PROJECTILE if (x, y) in rounded_coord else ' ' for x in range(columns)] for y in range(rows, -1, -1)]
```

Comment:
```py
if rows > 24 or columns > 55:
            return '\n Graph is too wide.'
```
Should we add something like that? Please note that these values are not accurate.

Reply:
yes, something like that. I don't know which values tho